bridges/centralized-ethereum/src/actors/dr_reporter.rs

```diff
@@ -133,451 +140,476 @@ impl Handler<DrReporterMsg> for DrReporter {
-    eth_account: H160,
-    report_result_limit: Option<u64>,
-    verbose: bool,
-    max_gas_price: U256,
+    eth_from: H160,
+    eth_gas_price: U256,
+    eth_nanowit_wei_price: U256,
+    eth_max_gas: Option<u64>,
 ) -> Vec<(Vec<Token>, U256)> {
     let mut v = vec![];
-    let mut stack = vec![batch_param];
+    let mut stack = vec![batched_report];
+
+    while let Some(batch_params) = stack.pop() {
+        let eth_report_result_batch_params = batch_params.clone();
+
+        // --------------------------------------------------------------------------
+        // First: try to estimate gas required for reporting this batch of tuples ...
 
-    while let Some(batch_param) = stack.pop() {
-        let params = (batch_param.clone(), verbose);
         let estimated_gas = wrb_contract
             .estimate_gas(
                 "reportResultBatch",
-                params,
-                eth_account,
+                eth_report_result_batch_params.clone(),
+                eth_from,
                 contract::Options::with(|opt| {
-                    opt.gas = report_result_limit.map(Into::into);
-                    opt.gas_price = Some(max_gas_price);
+                    opt.gas = eth_max_gas.map(Into::into);
+                    opt.gas_price = Some(eth_gas_price);
                 }),
             )
             .await;
+
+        if let Err(e) = estimated_gas {
+            if batch_params.len() <= 1 {
+                // Skip this single-query batch if still not possible to estimate gas
+                log::error!("Cannot estimate gas limit: {:?}", e);
+                log::warn!("Skipping report batch: {:?}", batch_params);
+            } else {
+                // Split batch in half if gas estimation is not possible
+                let (batch_tuples_1, batch_tuples_2) =
+                    batch_params.split_at(batch_params.len() / 2);
+                stack.push(batch_tuples_1.to_vec());
+                stack.push(batch_tuples_2.to_vec());
+            }
+
+            continue;
+        }
+
+        let estimated_gas = estimated_gas.unwrap();
         log::debug!(
-            "reportResultBatch {} estimated gas: {:?}",
-            batch_param.len(),
+            "reportResultBatch (x{} drs) estimated gas: {:?}",
+            batch_params.len(),
             estimated_gas
         );
 
-        match estimated_gas {
-            Ok(estimated_gas) => {
-                v.push((batch_param, estimated_gas));
+        // ------------------------------------------------
+        // Second: try to estimate actual profit, if any...
+
+        let query_ids: Vec<Token> = batch_params
+            .iter()
+            .map(|report_params| {
+                if let Token::Tuple(report_params) = report_params {
+                    assert_eq!(report_params.len(), 4);
+
+                    report_params[0].clone()
+                } else {
+                    panic!("Cannot extract query id from batch tuple");
```

Comment:
I'd try to avoid the use of `panic`s to avoid stopping the process. Could we use `log::error!` instead?

Reply:
Hmm, if this point is reached, it would imply that the bridge is interacting with a non-compatible contract, so there'd be no point to continue the process.